src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java

```diff
@@ -998,4 +1032,293 @@ private static MethodHandle stringValueOf(Class<?> ptype) {
+                break;
+            }
+
+            Class<?> ttype = ttypes[pos];
+            MethodHandle prepender = prepender(lastFragment.isEmpty() ? null : fragment, ttype);
+            initialLengthCoder = JLA.stringConcatMix(initialLengthCoder, fragment);
+            mh = MethodHandles.filterArgumentsWithCombiner(mh, 1, prepender,1, 0, 2 + pos);
+
+            pos++;
+        }
+
+        MethodHandle newArrayCombinator = lastFragment.isEmpty() ? newArray() :
+                newArrayWithSuffix(lastFragment);
+        mh = MethodHandles.foldArgumentsWithCombiner(mh, 0, newArrayCombinator,
+                1 // index
+        );
+
+        pos = 0;
+        for (Class<?> ttype : ttypes) {
+            MethodHandle mix = mixer(ttypes[pos]);
+            boolean lastPType = pos == ttypes.length - 1;
+
+            if (lastPType) {
+                mix = MethodHandles.insertArguments(mix, 0, initialLengthCoder);
+                mh = MethodHandles.foldArgumentsWithCombiner(mh, 0, mix,
+                        1 + pos // selected argument
+                );
+            } else {
+                mh = MethodHandles.filterArgumentsWithCombiner(mh, 0, mix,
+                        0, // old-index
+                        1 + pos // selected argument
+                );
+            }
+
+            pos++;
+        }
+
+        mh = MethodHandles.filterArguments(mh, 0, filters);
+        MethodType mt = MethodType.methodType(String.class, ptypes);
+        mh = mh.viewAsType(mt, true);
+
+        return mh;
+    }
+
+    /**
+     * This method breaks up large concatenations into separate
+     * {@link MethodHandle MethodHandles} based on the number of slots required
+     * per {@link MethodHandle}. Each {@link MethodHandle} after the first will
+     * have an extra {@link String} slot for the result from the previous
+     * {@link MethodHandle}.
+     * {@link java.lang.invoke.StringConcatFactory#makeConcatWithTemplate}
+     * is used to construct the {@link MethodHandle MethodHandles}. The total
+     * number of slots used by the ptypes is open ended. However, care must
+     * be given when combining the {@link MethodHandle MethodHandles} so that
+     * the combine total does not exceed the 255 slot limit.
+     *
+     * @param fragments list of string fragments
+     * @param ptypes    list of expression types
+     * @param maxSlots  maximum number of slots per {@link MethodHandle}.
+     *
+     * @return List of {@link MethodHandle MethodHandles}
+     *
+     * @throws IllegalArgumentException If maxSlots is not between 1 and
+     *                                  MAX_INDY_CONCAT_ARG_SLOTS.
+     * @throws StringConcatException If any of the linkage invariants are violated.
+     * @throws NullPointerException If any of the incoming arguments is null.
+     *
+     * @since 20
+     */
+    @PreviewFeature(feature=PreviewFeature.Feature.STRING_TEMPLATES)
+    public static List<MethodHandle> makeConcatWithTemplateCluster(
```

Comment:
I think instead of having two public methods and the user having to choose between them, it's better to have the implementations private and on public methods that calls the correct implementation if maxSlots > MAX_INDY_CONCAT_ARG_SLOTS or not

Reply:
Use cases are very different.  The first one produces a `MethodHandle` that has multiple inputs, The second one produces a `MethodHandle` that can only have one input.

Reply:
yes, sorry for the noise.